### PR TITLE
[1LP][RFR] Test targeted refresh for Nuage entities

### PR DESCRIPTION
**Pull request by sasoc**
With this commit we create fixture that tails 'evm.log' to see whether targeted refresh was triggered. We then use it to test targeted refresh for Nuage.

Fixture first starts tailing the log and waits for events to occur. Then it searches the 'evm.log' for a line that looks something like this:
```
Collection of targets with id: [{:ems_ref=>"b35d3afe-6f19-4da8-b1ee-b79de433cace"}, ... ]
```
If any of the log entry is not found `NeedleNotFoundInLog` error is raised and failed entities are listed.

NOTE: The fixture actually checks whether targeted refresh was _started_ i.e. whether Automate Instances are properly configured. It does not check the result of refreshing

Depends on:  https://github.com/ManageIQ/integration_tests/pull/7797, https://github.com/ManageIQ/integration_tests/pull/7812, https://github.com/ManageIQ/integration_tests/pull/7833

### CODE REVIEW

**Comment by dajoRH** [comment]
Would you mind rebasing this Pull Request against latest master, please? :trollface:
*CFME QE Bot*

**Comment by sasoc** [comment]
cc @miha-plesko

File: cfme/utils/appliance/__init__.py [hunk 1]
```diff
@@ -922,6 +922,10 @@ def event_listener(self):
             from cfme.utils.events import RestEventListener
             return RestEventListener(self)
 
+    def targeted_refresh_checker(self):
+        from cfme.utils.targeted_refresh import TargetedRefresh
+        return TargetedRefresh()
```
Comment (quarckster) [inline cfme/utils/appliance/__init__.py]: **QUESTION**
Why did you add it to the appliance object?

File: cfme/utils/targeted_refresh.py [hunk 2]
```diff
@@ -0,0 +1,51 @@
+# -*- coding: utf-8 -*-
+"""
+clas for targeted refresh tests
+"""
+
+
+class TargetedRefresh:
```
Comment (quarckster) [inline cfme/utils/targeted_refresh.py]: **REQUIRED**
```python
class TargetedRefresh(object):
```

File: cfme/utils/targeted_refresh.py [hunk 3]
```diff
@@ -0,0 +1,51 @@
...
+            if success:
+                self._triggered_refreshes.append(entity)
+
+    def new_targeted_refresh_check_string(self, data):
+        """
+        creates new regex with ems_ref to check logs
+        """
+        self._check_strings[data['entity']] = r'^.*refresh_targets_for_ems.*Collection ' \
+                                              r'of targets with id.*:ems_ref=>' \
+                                              r'"{}".*.$'.format(data['ems_ref'])
```
Comment (quarckster) [inline cfme/utils/targeted_refresh.py]: **REQUIRED**
```python
self._check_strings[data['entity']] = (r'^.*refresh_targets_for_ems.*Collection '
                                       r'of targets with id.*:ems_ref=>'
                                       r'"{}".*.$'.format(data['ems_ref']))
```

File: cfme/utils/targeted_refresh.py [hunk 4]
```diff
@@ -0,0 +1,51 @@
...
+        self._entities.append(data['entity'])
+
+    def get_entities(self):
```
Comment (quarckster) [inline cfme/utils/targeted_refresh.py]: **REQUIRED**
This should be a property.

File: cfme/utils/targeted_refresh.py [hunk 5]
```diff
@@ -0,0 +1,51 @@
...
+        return self._entities
+
+    def get_triggered_refreshes(self):
```
Comment (quarckster) [inline cfme/utils/targeted_refresh.py]: **REQUIRED**
^

File: cfme/utils/targeted_refresh.py [hunk 6]
```diff
@@ -0,0 +1,51 @@
+# -*- coding: utf-8 -*-
+"""
+clas for targeted refresh tests
```
Comment (quarckster) [inline cfme/utils/targeted_refresh.py]: typo

File: cfme/utils/appliance/__init__.py (hunk 1 above)
Comment (sasoc) [reply cfme/utils/appliance/__init__.py]: I used events as example. I've changed it, is it ok now?

**Comment by sasoc** [comment]
@quarckster I've implemented changes that you requested! Would you please take another look if its ok. Thank you for your review.

File: cfme/fixtures/merkyl.py [hunk 7]
```diff
@@ -68,6 +69,33 @@ def search_log(self, needle, log_name):
         else:
             return False
 
+    def search_log_with_regex(self, expr, log_name, _flag):
```
Comment (RonnyPfannschmidt) [inline cfme/fixtures/merkyl.py]: **REQUIRED** use actual regex flags instea of magical constants

File: cfme/utils/targeted_refresh.py [hunk 8]
```diff
@@ -0,0 +1,53 @@
...
+                'MULTILINE'
+            )
+            if success:
+                self._triggered_refreshes.append(entity)
+
+    def new_targeted_refresh_check_string(self, data):
+        """
+        creates new regex with ems_ref to check logs
+        """
+        self._check_strings[data['entity']] = (r'^.*brerefresh_targets_for_ems.*Collection '
```
Comment (RonnyPfannschmidt) [inline cfme/utils/targeted_refresh.py]: **question** why not put in a compiled regex with the correct flags here

File: cfme/utils/targeted_refresh.py [hunk 9]
```diff
@@ -0,0 +1,53 @@
...
+
+class TargetedRefresh(object):
+
+    def __init__(self):
+        self._check_strings = {}
+        self._entities = []
+        self._triggered_refreshes = []
+
+    @property
+    def entities(self):
```
Comment (RonnyPfannschmidt) [inline cfme/utils/targeted_refresh.py]: **required** those properties offer no value in python, please use direct attributes

File: cfme/utils/targeted_refresh.py [hunk 10]
```diff
@@ -0,0 +1,53 @@
+# -*- coding: utf-8 -*-
+"""
+class for targeted refresh tests
+"""
+
+
+class TargetedRefresh(object):
+
+    def __init__(self):
```
Comment (RonnyPfannschmidt) [inline cfme/utils/targeted_refresh.py]: **optional** use attrs to declare the class

File: cfme/networks/provider/__init__.py [hunk 11]
```diff
@@ -96,7 +96,11 @@ def routers(self):
     def security_groups(self):
         return self.collections.security_groups
 
-    def create(self, cancel=False, validate_credentials=True, validate_inventory=False):
+    def create(self, cancel=False, validate_credentials=True, validate_inventory=False,
+               check_existing=False):
+        if check_existing and self.exists:
```
Comment (RonnyPfannschmidt) [inline cfme/networks/provider/__init__.py]: @mshriver please take a look here - i believe its conceptually ok but i wonder if the flag should be called `exists_ok` or if it should be a separate method

File: cfme/fixtures/artifactor_plugin.py [hunk 12]
```diff
@@ -150,12 +150,12 @@ def fire_art_hook(config, hook, **hook_args):
     if client is None:
         assert UNDER_TEST, 'missing artifactor is only valid for inprocess tests'
     else:
-        client.fire_hook(hook, **hook_args)
+        return client.fire_hook(hook, **hook_args)
```
Comment (RonnyPfannschmidt) [inline cfme/fixtures/artifactor_plugin.py]: **question** where are those used - i believe i'm missing a link in chain there

**Comment by dajoRH** [comment]
Would you mind rebasing this Pull Request against latest master, please? :trollface:
*CFME QE Bot*

**Comment by dajoRH** [comment]
Lint report for commit b38ff32e92f05af9c0818bc3c6158b6fd949d5d1:

`cfme/tests/networks/nuage/test_events_are_triggered.py`:
- :bangbang: Line 130:101: **E501** *line too long (101 > 100 characters)*

Please, rectify these issues :smirk: .


**No commit flaws detected**.

*CFME QE Bot*

File: cfme/fixtures/events.py [hunk 13]
```diff
@@ -81,6 +81,24 @@ def pytest_runtest_call(item):
                                 "Event {} did not come!".format(event['event']))
             else:
                 logger.info('Seems like all events have arrived!')
+        elif "register_targeted_refresh" in item.funcargnames:
```
Comment (miha-plesko) [inline cfme/fixtures/events.py]: So basically we need to wait some time after main code block (of the test) executes and check if specific lines have appeared in the evm.log. But do we really need to hook this here? Couldn't we just do it a fixture after the `yield`? 

@izapolsk maybe ^, since you've been involved in event listener.
@sasoc have you tried it and there was an error?

File: cfme/fixtures/merkyl.py (hunk 7 above)
Comment (sasoc) [reply cfme/fixtures/merkyl.py]: Done.

File: cfme/utils/targeted_refresh.py (hunk 9 above)
Comment (sasoc) [reply cfme/utils/targeted_refresh.py]: Done.

File: cfme/utils/targeted_refresh.py (hunk 10 above)
Comment (sasoc) [reply cfme/utils/targeted_refresh.py]: Done.

File: cfme/networks/provider/__init__.py (hunk 11 above)
Comment (sasoc) [reply cfme/networks/provider/__init__.py]: This was already done in some other PR now.

File: cfme/fixtures/artifactor_plugin.py (hunk 12 above)
Comment (sasoc) [reply cfme/fixtures/artifactor_plugin.py]: This is now in a different PR(https://github.com/ManageIQ/integration_tests/pull/7763). Where I commented(https://github.com/ManageIQ/integration_tests/pull/7763#issuecomment-418812041) why this needs to return value.

File: cfme/fixtures/events.py (hunk 13 above)
Comment (miha-plesko) [reply cfme/fixtures/events.py]: In any case @sasoc do we really want to extend their function here? I'd rather implement our own hook function in the same file that our fixture is defined...

File: cfme/fixtures/targeted_refresh.py [hunk 14]
```diff
@@ -0,0 +1,32 @@
+"""
+The idea of this fixture is to pass some entity ids to
+:py:class:`utils.targeted_refresh.TargetedRefresh` and check the evm.log
+whether all expected targeted refreshes
+were triggered.
+
+register_targeted_refresh accepts attributes for entity that is expected to trigger refresh
+"""
```
Comment (miha-plesko) [inline cfme/fixtures/targeted_refresh.py]: This description literally tells nothing to me. I'd expect to get high-level overview of the thing first and then some usage examples to get me started with.

File: cfme/utils/targeted_refresh.py [hunk 15]
```diff
@@ -0,0 +1,48 @@
...
+class for targeted refresh tests
+"""
+
+import re
+import attr
+
+
+@attr.s
+class TargetedRefresh(object):
+
```
Comment (miha-plesko) [inline cfme/utils/targeted_refresh.py]: Please remove the empty line here

File: cfme/fixtures/events.py (hunk 13 above)
Comment (sasoc) [reply cfme/fixtures/events.py]: @miha-plesko I will implement it in my own hook function.

Comment (miha-plesko) [reply cfme/fixtures/events.py]: @sasoc can you please try to implement everything as a normal fixture without the hook function instead? As far as I can understand the hook function is only needed because of the soft_assert that you don't really need IMAO.

File: cfme/tests/networks/nuage/test_events_are_triggered.py [hunk 16]
```diff
@@ -120,3 +120,23 @@ def cmp_function(_, full_data):
         event_type=event_type,
         first_event=True
     )
+
+
+def test_targeted_refresh_after_create_entities(register_targeted_refresh, with_nuage_sandbox):
+    """
+    Test whether targeted refresh is triggered when entities are created.
+
+    config:
```
Comment (miha-plesko) [inline cfme/tests/networks/nuage/test_events_are_triggered.py]: Prerequisities:

File: cfme/tests/networks/nuage/test_events_are_triggered.py [hunk 17]
```diff
@@ -120,3 +120,23 @@ def cmp_function(_, full_data):
...
+        Before running this test you have to configure artifactor(see how in
```
Comment (miha-plesko) [inline cfme/tests/networks/nuage/test_events_are_triggered.py]: space between `artifactor(see`

File: cfme/fixtures/events.py (hunk 13 above)
Comment (sasoc) [reply cfme/fixtures/events.py]: @miha-plesko Do you have an advice what to do in case of multiple entities to check, if I use normal fixture? For example if we check for subnet, enterprise and zone. What to do in case that 2 of them don't trigger and one does. How and when to display an error?

File: cfme/fixtures/targeted_refresh.py [hunk 18]
```diff
@@ -0,0 +1,32 @@
...
+       This fixture is used to notify the testing system that targeted refresh should
+       be triggered and that conformation should be fount in evm.log
+
+       Returns: None
+
+       Usage:
+
+           def test_something(foo, bar, register_targeted_refresh, appliance, merkyl_inspector):
+               merkyl_inspector.add_log('path/to/log/name.log')
+               register_targeted_refresh(entity = 'subnet', ems_ref = subnet.ems_ref)
```
Comment (miha-plesko) [inline cfme/fixtures/targeted_refresh.py]: Please remove spaces around `=`

File: cfme/fixtures/events.py (hunk 13 above)
Comment (miha-plesko) [reply cfme/fixtures/events.py]: Concat errors in a single string and raise it with hard assert. I mean, doing all this woodoo magic just to call soft_assert sounds like a nonsense to me since we have a list of all the found matches at hand all the time.

File: cfme/tests/networks/nuage/test_events_are_triggered.py [hunk 19]
```diff
@@ -120,3 +120,23 @@ def cmp_function(_, full_data):
...
+
+    Steps:
+        * Deploy entities outside of CFME (directly in the provider)
+        * Verify that targeted refreshes where triggered
+    """
+    entities_to_test = ['enterprise']
+    sandbox = with_nuage_sandbox
+
+    for entity in entities_to_test:
+        register_targeted_refresh(entity=entity, ems_ref=sandbox['enterprise'].id)
```
Comment (miha-plesko) [inline cfme/tests/networks/nuage/test_events_are_triggered.py]: `'enterprise'` should not be hardcoded here

File: cfme/fixtures/events.py (hunk 13 above)
Comment (sasoc) [reply cfme/fixtures/events.py]: @miha-plesko Ok thank you for the advice.

File: cfme/fixtures/targeted_refresh.py [hunk 20]
```diff
@@ -0,0 +1,32 @@
...
+from cfme.utils.targeted_refresh import TargetedRefresh
+
+
+@pytest.fixture(scope="function")
+def register_targeted_refresh(soft_assert, merkyl_inspector):
+    """register_targeted_refresh(entity, entity.ems_ref, )
+       register targeted fixture.
+
+       This fixture is used to notify the testing system that targeted refresh should
+       be triggered and that conformation should be fount in evm.log
```
Comment (miha-plesko) [inline cfme/fixtures/targeted_refresh.py]: conformation -> log entry
fount -> found

File: cfme/fixtures/targeted_refresh.py [hunk 21]
```diff
@@ -0,0 +1,32 @@
...
+
+       Returns: None
+
+       Usage:
+
+           def test_something(foo, bar, register_targeted_refresh, appliance, merkyl_inspector):
+               merkyl_inspector.add_log('path/to/log/name.log')
```
Comment (miha-plesko) [inline cfme/fixtures/targeted_refresh.py]: Doesn't fixture take care of this?

File: cfme/tests/networks/nuage/test_events_are_triggered.py (hunk 19 above)
Comment (miha-plesko) [reply cfme/tests/networks/nuage/test_events_are_triggered.py]: Nevertheless, this won't work as sandbox keys are not always entity names (look at `cont_vport`, for example). I'd suggest to just do it explicitly for each event we want to capture

```
register_targeted_refresh('enterprise', sandbox['enterprise'].id)
```

File: cfme/utils/targeted_refresh.py [hunk 22]
```diff
@@ -0,0 +1,48 @@
...
+    check_strings = attr.ib(default={})
+    entities = attr.ib(default=[])
+    triggered_refreshes = attr.ib(default=[])
```
Comment (miha-plesko) [inline cfme/utils/targeted_refresh.py]: Two concerns here:

a) setting default value to mutable object and then modify it is never a good idea because strange behavior may occur, see https://stackoverflow.com/questions/44141104/when-and-why-should-i-use-attr-factory
b) why would you even want to let caller set those internal variables? Please consider marking them read-only like this:

```python
_triggered_refreshes = attr.ib(init=False, default=attr.Factory(list))
```

See http://www.attrs.org/en/stable/examples.html

File: cfme/fixtures/targeted_refresh.py [hunk 23]
```diff
@@ -0,0 +1,32 @@
...
+were triggered.
+
+register_targeted_refresh accepts attributes for entity that is expected to trigger refresh
+"""
+import pytest
+from cfme.utils.targeted_refresh import TargetedRefresh
+
+
+@pytest.fixture(scope="function")
+def register_targeted_refresh(soft_assert, merkyl_inspector):
```
Comment (miha-plesko) [inline cfme/fixtures/targeted_refresh.py]: Can we change fixture name to `targeted_refresh` and make it return object with method `register_target`? I would find it more readable in test to say:

```
targeted_refresh.register_target(...)
targeted_refresh.register_target(...)
targeted_refresh.register_target(...)
```

There is also a comment in register_event suggesting the same, saying literally `bad idea, to replace by object later`, see https://github.com/ManageIQ/integration_tests/blob/master/cfme/utils/events.py#L283

File: cfme/utils/targeted_refresh.py [hunk 24]
```diff
@@ -0,0 +1,48 @@
...
+"""
+class for targeted refresh tests
+"""
+
+import re
+import attr
+
+
+@attr.s
+class TargetedRefresh(object):
```
Comment (miha-plesko) [inline cfme/utils/targeted_refresh.py]: What this class in fact it does is it checks specific log file and searches for a list of regular expressions to occur until stopped or until all regexes are found, with some timeout.

This is very useful feature IMO and we should make it reusable for arbitrary regular expressions, not lock it for targeted refresh only. Therefore I suggest we rename the class to

```
class LogFinder:
   ...
```

and allow the caller to use it for searching arbitrary regular expressions. We would then modify the caller, which is `targeted_refresh` fixture in our case, to form the correct regex, which is 

```python
r'^.*refresh_targets_for_ems.*Collection of targets with id.*:ems_ref=> "{}".*.$'.format('ems_ref')
```

in our case.

Wdyt?

Comment (sasoc) [reply cfme/utils/targeted_refresh.py]: This is a great idea in IMO. I already thought about making it reusable for other cases but didn't know how. I think I will make a new PR for this.

**Comment by sasoc** [comment]
@miha-plesko I've simplified this test just like we talked about in person. Would you please take a look if its like you imagined.

File: cfme/tests/networks/nuage/test_events_are_triggered.py [hunk 25]
```diff
@@ -120,3 +124,98 @@ def cmp_function(_, full_data):
...
+        yield
+
+        def check_log():
+            for target in set(targets):
+                success = merkyl_inspector.search_log(
+                    target[0],
+                    '/var/www/miq/vmdb/log/evm.log'
+                )
+
+                if success:
```
Comment (miha-plesko) [inline cfme/tests/networks/nuage/test_events_are_triggered.py]: Do we really need intermediate variable?

```python
if merkyl_inspector.search_log(...):
  targets.remove(target)
```

File: cfme/tests/networks/nuage/test_events_are_triggered.py [hunk 26]
```diff
@@ -120,3 +124,98 @@ def cmp_function(_, full_data):
...
+    Usage:
+
+        def test_something(targeted_refresh)
+                trigger_targeted_refresh() <- some function that will trigger
+                                              targeted refresh
+                with targeted_refresh.target_timeout():
+                    targeted_refresh.register_target(domain_id, 'domain')
+                    targeted_refresh.register_target(cont_vport, 'vport', comment='L3 cont')
+    """
+    merkyl_inspector.add_log('/var/www/miq/vmdb/log/evm.log')
```
Comment (miha-plesko) [inline cfme/tests/networks/nuage/test_events_are_triggered.py]: Initailaize variable for evm.log path because we use it multiple times:

```python
evm_log = '/var/www/miq/vmdb/log/evm.log'
```

File: cfme/tests/networks/nuage/test_events_are_triggered.py [hunk 27]
```diff
@@ -120,3 +124,98 @@ def cmp_function(_, full_data):
...
+    targets = set()
+
+    @contextmanager
+    def target_timeout():
+        yield
+
+        def check_log():
```
Comment (miha-plesko) [inline cfme/tests/networks/nuage/test_events_are_triggered.py]: Is there a reason why check_log needs to be nested in `timeout()` function? Couldn't we just move it out into fixture scope directly?

File: cfme/tests/networks/nuage/test_events_are_triggered.py [hunk 28]
```diff
@@ -120,3 +124,98 @@ def cmp_function(_, full_data):
...
+            for target in set(targets):
+                success = merkyl_inspector.search_log(
+                    target[0],
+                    '/var/www/miq/vmdb/log/evm.log'
```
Comment (miha-plesko) [inline cfme/tests/networks/nuage/test_events_are_triggered.py]: Use `evm_log` variable

File: cfme/tests/networks/nuage/test_events_are_triggered.py [hunk 29]
```diff
@@ -120,3 +124,98 @@ def cmp_function(_, full_data):
...
+
+        def test_something(targeted_refresh)
+                trigger_targeted_refresh() <- some function that will trigger
+                                              targeted refresh
+                with targeted_refresh.target_timeout():
+                    targeted_refresh.register_target(domain_id, 'domain')
+                    targeted_refresh.register_target(cont_vport, 'vport', comment='L3 cont')
+    """
+    merkyl_inspector.add_log('/var/www/miq/vmdb/log/evm.log')
+    targets = set()
```
Comment (miha-plesko) [inline cfme/tests/networks/nuage/test_events_are_triggered.py]: Please add inline comment like this to document what we're going to put in the set

```python
targets = set()  # { (ems_ref, comment), (ems_ref, comment), ... }
```

File: cfme/tests/networks/nuage/test_events_are_triggered.py [hunk 30]
```diff
@@ -120,3 +124,98 @@ def cmp_function(_, full_data):
...
+
+        try:
+            wait_for(check_log,
+                     delay=5,
+                     num_sec=180,
+                     handle_exception=True)
+        except TimedOutError:
+            raise MatchNotFoundInLog('Targeted refresh did not trigger for ({})'.format(
+                ', '.join(list(map(lambda t: t[2] or t[1], targets)))))
+        else:
```
Comment (miha-plesko) [inline cfme/tests/networks/nuage/test_events_are_triggered.py]: No need for this else

File: cfme/tests/networks/nuage/test_events_are_triggered.py [hunk 31]
```diff
@@ -120,3 +124,98 @@ def cmp_function(_, full_data):
...
+
+                if success:
+                    targets.remove(target)
+            return len(targets) == 0
+
+        try:
+            wait_for(check_log,
+                     delay=5,
+                     num_sec=180,
+                     handle_exception=True)
```
Comment (miha-plesko) [inline cfme/tests/networks/nuage/test_events_are_triggered.py]: Not sure, but shouldn't this be set to False? I understand we now suppress all exceptions that may occur in `check_log` which I think is not desired. Please just delete the argument, False is default I see.

File: cfme/tests/networks/nuage/test_events_are_triggered.py [hunk 32]
```diff
@@ -120,3 +124,98 @@ def cmp_function(_, full_data):
         event_type=event_type,
         first_event=True
     )
+
+
+@pytest.fixture
+def targeted_refresh(merkyl_inspector):
```
[the remark this answers is not in the document]
Comment (miha-plesko) [reply cfme/tests/networks/nuage/test_events_are_triggered.py]: Also: do we need to list `merkyl_setup` here as well or not? What does it do anyway?

File: cfme/tests/networks/nuage/test_events_are_triggered.py [hunk 33]
```diff
@@ -120,3 +124,98 @@ def cmp_function(_, full_data):
...
+            return len(targets) == 0
+
+        try:
+            wait_for(check_log,
+                     delay=5,
+                     num_sec=180,
+                     handle_exception=True)
+        except TimedOutError:
+            raise MatchNotFoundInLog('Targeted refresh did not trigger for ({})'.format(
+                ', '.join(list(map(lambda t: t[2] or t[1], targets)))))
```
Comment (miha-plesko) [inline cfme/tests/networks/nuage/test_events_are_triggered.py]: Do you really need the outer `list()`? I think `map` returns list of string already

Comment (miha-plesko) [reply cfme/tests/networks/nuage/test_events_are_triggered.py]: Also, we should break line after each target desription because we don't want them all in a single line...

File: cfme/tests/networks/nuage/test_events_are_triggered.py (hunk 32 above)
[the remark this answers is not in the document]
Comment (sasoc) [reply cfme/tests/networks/nuage/test_events_are_triggered.py]: If I understand correctly,`merkyl_setup` is used to add log files at the start of the session. For that you need to list them in configuration under log_files (https://github.com/ManageIQ/integration_tests/blob/master/artifactor/plugins/merkyl.py#L13-L16)

But in our case we use Merkyl for the duration of the test and we add log with `merkyl_inspector.add()`

[the remark this answers is not in the document]
Comment (miha-plesko) [reply cfme/tests/networks/nuage/test_events_are_triggered.py]: Thanks for explanation. I agree with you, we don't need it then 👍

**Comment by sasoc** [comment]
@miha-plesko Thank you for your review. I've commited the changes you requested.

File: cfme/tests/networks/nuage/test_events_are_triggered.py [hunk 34]
```diff
@@ -95,6 +151,38 @@ def test_creating_entities_emits_events(register_event, with_nuage_sandbox):
...
+        targeted_refresh.register_target(sandbox['enterprise'].id, 'Enterprise')
+        targeted_refresh.register_target(sandbox['domain'].id, 'Domain')
+        targeted_refresh.register_target(sandbox['subnet'].id, 'Subnet')
+        targeted_refresh.register_target(sandbox['cont_vport'].id, 'Container vPort (L3)')
+        targeted_refresh.register_target(sandbox['vm_vport'].id, 'VM vPort (L3)')
+        targeted_refresh.register_target(sandbox['l2_domain'].id, 'Domain (L2)')
+        targeted_refresh.register_target(sandbox['l2_cont_vport'].id, 'Container vPort (L2)')
+        targeted_refresh.register_target(sandbox['l2_vm_vport'].id, 'VM vPort (L2)')
+        targeted_refresh.register_target(sandbox['group'].id, 'L3')
+        targeted_refresh.register_target(sandbox['l2_group'].id, 'L2')
```
Comment (miha-plesko) [inline cfme/tests/networks/nuage/test_events_are_triggered.py]: Fix the last tow comments, please:
'L3' -> 'Security group (L3)'
'L2' -> 'Security group (L2)'

**Comment by sasoc** [comment]
@miha-plesko I've pushed the latest changes you requested. Once again thank you for the review.

File: cfme/tests/networks/nuage/test_events_are_triggered.py [hunk 35]
```diff
@@ -61,6 +65,53 @@ def with_nuage_sandbox(networks_provider):
...
+                    targeted_refresh.register_target('ref-123456', 'Subnet named TEST')
+                    targeted_refresh.register_target('ref-aaaabb', 'Router named TEST')
+    """
+    evm_log = '/var/www/miq/vmdb/log/evm.log'
+    needle_template = r'^.*Collection of targets with id.*:ems_ref=>"{}".*.$'
+    merkyl_inspector.add_log(evm_log)
+    targets = set()  # { (ems_ref, comment), (ems_ref, comment), ... }
+
+    def check_log():
+        for target in set(targets):
```
Comment (miha-plesko) [inline cfme/tests/networks/nuage/test_events_are_triggered.py]: Please add following line right before the `for` loop

```
logger.info('Looking for %s needles in evm.log: %s', len(targets), [t[1] for t in targets])
```

in order for user to be able to follow the progress when the test runs.

File: cfme/tests/networks/nuage/test_events_are_triggered.py [hunk 36]
```diff
@@ -61,6 +65,53 @@ def with_nuage_sandbox(networks_provider):
...
+            if merkyl_inspector.search_log(target[0], evm_log):
+                targets.remove(target)
```
Comment (miha-plesko) [inline cfme/tests/networks/nuage/test_events_are_triggered.py]: Please add following line prior removing target

```
logger.info('Found needle %s', target[1])
```

File: cfme/tests/networks/nuage/test_events_are_triggered.py [hunk 37]
```diff
@@ -61,6 +65,53 @@ def with_nuage_sandbox(networks_provider):
...
+                with targeted_refresh.target_timeout():
+                    targeted_refresh.register_target('ref-123456', 'Subnet named TEST')
+                    targeted_refresh.register_target('ref-aaaabb', 'Router named TEST')
+    """
+    evm_log = '/var/www/miq/vmdb/log/evm.log'
+    needle_template = r'^.*Collection of targets with id.*:ems_ref=>"{}".*.$'
+    merkyl_inspector.add_log(evm_log)
+    targets = set()  # { (ems_ref, comment), (ems_ref, comment), ... }
+
+    def check_log():
```
Comment (miha-plesko) [inline cfme/tests/networks/nuage/test_events_are_triggered.py]: So what check_log currently does is it fetches evm.log content from the appliance for each of the 10 targets in our test and attempts to find needle in it. This means we transfer the evm.log content 10x every 5 seconds! Please consider refactoring this function to only fetch the evm.log once every 5 seconds, for example:

```python
def check_log():
        logger.info('Looking for %s needles in evm.log: %s', len(targets), [t[1] for t in targets])
        content = merkyl_inspector.get_log(evm_log)
        for target in set(targets):
            if target[0].search(content):
                logger.info('Found needle %s', target[1])
                targets.remove(target)
        return len(targets) == 0
```

File: cfme/tests/networks/nuage/test_events_are_triggered.py [hunk 38]
```diff
@@ -61,6 +65,53 @@ def with_nuage_sandbox(networks_provider):
     logger.info('Destroyed sandbox enterprise {} ({})'.format(enterprise.name, enterprise.id))
 
 
+@pytest.fixture
+def targeted_refresh(merkyl_inspector):
```
Comment (miha-plesko) [inline cfme/tests/networks/nuage/test_events_are_triggered.py]: This does not work for me unless I require `merkyl_setup` fixture here as well. What happens otherwise is that accessing `/get/evm.log` yields following error:

```
<!DOCTYPE HTML PUBLIC "-//IETF//DTD HTML 2.0//EN">
<html>
<head>
	<title>Error: 500 Internal Server Error</title>
</head>
<body>
	<h1>Error: 500 Internal Server Error</h1>
	<p>Sorry, the requested URL <tt>&#039;http://10.10.43.130:8192/get/evm.log&#039;</tt>
	   caused an error:</p>
	<pre>Internal Server Error</pre>
</body>
</html>
```

Not sure why, but if I add the merkyl_setup fixture it starts working. If I remove it, it stops working again. Do you happen to see a reason why would .add_log not be working?

Comment (sasoc) [reply cfme/tests/networks/nuage/test_events_are_triggered.py]: Is there any chance you have `evm.log` listed under log_files in configuration:
```
merkyl:
    log_files:
        - /var/www/miq/vmdb/log/evm.log
```
If this is the case `.add_log` skips (see here: https://github.com/ManageIQ/integration_tests/blob/e0e650c801c7eeb8f5c2230e9e13fda77e082e66/artifactor/plugins/merkyl.py#L79) adding the log, since it presumes that is was already created in merkyl_setup.

Comment (miha-plesko) [reply cfme/tests/networks/nuage/test_events_are_triggered.py]: Okay, so this problem is now settled: Merkyl won't add log if you have it listed in your env.local.yaml because it assumes it was already added upon setup _even_ if you didn't invoke the setup. The solution to this is to also require merkyl_setup so that the setup actually occurs - and then everything works.

@sasoc if user will have evm.log listed in env.local.yaml then it will be opened for entire session, which is a problem because our test will receive content from previous tests as well. So please make sure you reset the file.

File: cfme/exceptions.py [hunk 39]
```diff
@@ -390,3 +390,9 @@ def __init__(self, collection, filter_key):
     def __str__(self):
         return 'Action on Collection ({}) requires a filter: ({})'\
                .format(self.collection, self.filter_key)
+
+
+class NeedleNotFoundInLog(CFMEException):
+    """
+    Raised when log doesnt't contain needle
```
Comment (digitronik) [inline cfme/exceptions.py]: Required:
This should be an `inline` comment.

Comment (sasoc) [reply cfme/exceptions.py]: @digitronik Thank you for the review! I've changed it to inline.

File: cfme/tests/networks/nuage/test_events_are_triggered.py [hunk 40]
```diff
@@ -61,6 +65,57 @@ def with_nuage_sandbox(networks_provider):
...
+    evm_log = '/var/www/miq/vmdb/log/evm.log'
+    needle_template = r'^.*Collection of targets with id.*:ems_ref=>"{}".*$'
+    merkyl_inspector.add_log(evm_log)
+    merkyl_inspector.reset_log(evm_log)
+    targets = set()  # { (ems_ref, comment), (ems_ref, comment), ... }
+
+    def check_log():
+        logger.info('Looking for %s needles in evm.log: %s', len(targets), [t[1] for t in targets])
+        content = merkyl_inspector.get_log(evm_log)
+        for target in set(targets):
```
Comment (izapolsk) [inline cfme/tests/networks/nuage/test_events_are_triggered.py]: it is already set.

**Comment by izapolsk** [comment]
please remind me when https://github.com/ManageIQ/integration_tests/pull/7833  gets merged.
I'll merge this PR

**Comment by mshriver** [comment]
All dependent PRs have now been merged.

I'm curious if LogValidator was ever considered to implement this test:
https://github.com/ManageIQ/integration_tests/blob/master/cfme/utils/log_validator.py#L8

That said, the implementation is clean and concise, and included some enhancements to merkyl, so I'm going to merge it. Wanted to at least bring up LogValidator so you're aware of it, if you weren't already.

Thanks!

**Comment by miha-plesko** [comment]
Ha, thanks for bringing it up @mshriver, we weren't aware of it indeed. From what I understand, LogValidator is in fact preferred over Merkyl as all it needs is SSH access, no deamon on appliance.

@sasoc wdyt, should we refactor the fixture to use this new approach? I think it should be pretty straight-forward - exactly same fixture structure only a bit different function implementations. Test can even remain unchanged.

@mshriver we're thinking to move our Nuage fixtures into a separate file as we want to reuse them in multiple tests. We're thinking to put them in `tests/networks/nuage/fixtures.py`. Would this work for you guys or is there some better place for them?

**Comment by mshriver** [comment]
@miha-plesko indeed, no daemon required simplifies things and protects your tests from merkyl getting deprecated for some reason (hypothetical).

I would recommend creating a cfme/fixtures/networks.py for common network provider fixtures.  I just commented on another PR about how registering such a fixture module in the pytest_plugin module allows for the included fixtures to be used throughout the framework, without having to import it directly.

**Comment by miha-plesko** [comment]
Thanks, we appreciate your comments 👍 Will definetly put `targeted_refresh` fixture in cfme/fixtures as it's general. We have, however, a fixture that is 100% Nuage specific ([with_nuage_sandbox](https://github.com/ManageIQ/integration_tests/blob/master/cfme/tests/networks/nuage/test_events_are_triggered.py#L21)) and we would like to use it in multiple test files, so it feels like it should be put it in nuage directory, should you agree. Or we can create cfme/fixtures/nuage.py file and stuff those in there.
